### DISCUSSION TITLE
How to use Turso without embedded replicas?

### DISCUSSION THREAD

**Discussion by touhidurrr** [prisma/prisma#25931] [Q&A] [answered]
### Question

So, this is my schema:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["driverAdapters"]
}

datasource db {
  provider = "sqlite"
  url      = env("TURSO_DATABASE_URL")
}
```
This works with prisma generate and I can do the queries fine if the tables are already there. But when I want to migrate, this happens:
```ts
~$ bunx prisma migrate dev --name init
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": SQLite database

Error: Prisma schema validation - (get-config wasm)
Error code: P1012
error: Error validating datasource `db`: the URL must start with the protocol `file:`.
  -->  prisma\schema.prisma:14
   |
13 |   provider = "sqlite"
14 |   url      = env("TURSO_DATABASE_URL")        
   |

Validation Error Count: 1
[Context: getConfig]

Prisma CLI Version : 6.1.0
```
I want to use Turso with prisma but without embedded replicas. (As I use stateless servers with low disk limits). How can I do this?

**Comment by nurul3101** [comment]
Hello @touhidurrr 👋 

To make database changes you would need to use Turso CLI as described [here](https://www.prisma.io/docs/orm/overview/databases/turso#differences-to-consider).

Since libSQL uses HTTP to connect to the remote database, this makes it incompatible with Prisma Migrate, which is why you are getting this error. 

<img width="917" alt="Screenshot 2024-12-26 at 3 46 55 PM" src="https://github.com/user-attachments/assets/62760ca5-d5d1-4804-b6e1-bf65dffa1bcc" />

**Reply by touhidurrr** [reply]
@nurul3101 my goal is not to implement schema changes to the database straight from **prisma** here. I want `prisma migrate` to just emit the sql code in migration folder so that I can then feed it to the database. Note: that this is also what **prisma** documentation tells you to do now but the problem is to use `prisma migrate`, you have to have replicas enabled which is what I don't want it to use now. I want to use prisma just to query the Turso database and to emit sql code for me.

![image](https://github.com/user-attachments/assets/45d2525c-b150-4eb2-ac16-ae0e559bd09f)

**Reply by touhidurrr** [reply]
In short what I am saying is "Give me some way to use Turso without replicas". Note that enabling replica is option using their official client library. But then again I would have to write raw sql and thats why I am using Prisma here.

**Reply by touhidurrr** [reply]
One possible solution can be:
1. Change url validation to accept `libsql://` urls.
2. Add an additional field replica that accepts filepath to enable replicas.

**Reply by nurul3101** [reply, marked as answer]
> but the problem is to use prisma migrate, you have to have replicas enabled

Prisma Migrate is currently not supported when working with Turso. This is not due to a requirement for replicas, but rather because of how Turso connects to the database.

The recommended workflow is to generate a migration file using `prisma migrate dev` against a local SQLite database (Not Turso) and then Applying the migration manually using Turso's CLI (To Turso database).

> Change url validation to accept libsql:// urls. Add an additional field replica that accepts filepath to enable replicas.

I would recommend opening a [Feature Request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&projects=&template=feature_request.md&title=) so that our ORM team can share their insights.

**Reply by touhidurrr** [reply]
It seems like my struggles with prisma is not going to stop anytime soon. https://github.com/prisma/prisma/issues/25939 https://github.com/prisma/prisma/issues/25948 https://github.com/prisma/prisma/issues/25932
